File: training/python/canastra_train/evaluate.py
"""Duplicate-deal evaluation: same seeds, swapped seatings, score differentials.

Canastra's variance is dominated by the deal, so two genomes are compared by
playing each seed TWICE — once with genome A in seats 0/2, once with the
seats swapped — and averaging A's score differential over the pair. The deal
cancels; what remains is policy.
"""

from __future__ import annotations

from dataclasses import dataclass

import numpy as np
import torch
from canastra_py import Pool

from canastra_train import genome as genome_mod
from canastra_train import policy


@dataclass
class PairReport:
    pairs: int
    mean_diff: float
    ci95: float
    wins_a: int
    wins_b: int
    unfinished: int


def evaluate_pair(
    vec_a: np.ndarray,
    vec_b: np.ndarray,
    arch: genome_mod.Arch,
    seeds: list[int],
    cap: int = 200_000,
) -> PairReport:
    """A vs B over `seeds`, each seed played in both seatings."""
    if len(set(seeds)) != len(seeds):
        raise ValueError("seeds must be unique — the pairing groups by seed")
    count = len(seeds)
    pool_seeds = seeds + seeds  # first half: A in seats 0/2; second half: swapped
    pool = Pool(pool_seeds, max_actions_per_game=cap)
    trunk_a, head_a = genome_mod.to_modules(vec_a, arch)
    trunk_b, head_b = genome_mod.to_modules(vec_b, arch)

    while pool.has_live():
        obs, acts, mask, rows = pool.encode()
        games = rows[:, 0]
        seats = rows[:, 1]
        a_is_team_zero = games < count
        use_a = (seats % 2 == 0) == a_is_team_zero
        picks = np.empty(len(rows), dtype=np.int64)
        if use_a.any():
            sel = np.flatnonzero(use_a)
            scores_a = policy.logits(
                trunk_a, head_a,
                torch.from_numpy(obs[sel]),
                torch.from_numpy(acts[sel]),
                torch.from_numpy(mask[sel]),
            )
            picks[sel] = policy.pick_argmax(scores_a)
        if (~use_a).any():
            sel = np.flatnonzero(~use_a)
            scores_b = policy.logits(
                trunk_b, head_b,
                torch.from_numpy(obs[sel]),
                torch.from_numpy(acts[sel]),
                torch.from_numpy(mask[sel]),
            )
            picks[sel] = policy.pick_argmax(scores_b)
        pool.apply(picks.tolist())

    results = pool.results()
    assert len(results) == 2 * count

    diffs: list[float] = []
    unfinished = 0
    wins_a = wins_b = 0
    by_seed: dict[int, list[float]] = {}
    for index, (_seed, pair_scores, winner, _hands, is_unfinished) in enumerate(results):
        a_is_team_zero = index < count
        a_score = pair_scores[0] if a_is_team_zero else pair_scores[1]
        b_score = pair_scores[1] if a_is_team_zero else pair_scores[0]
        by_seed.setdefault(_seed, []).append(a_score - b_score)
        if is_unfinished:
            unfinished += 1
        elif winner is not None:
            won_a = (winner == 0) == a_is_team_zero
            if won_a:
                wins_a += 1
            else:
                wins_b += 1

    for seed in seeds:
        pair = by_seed[seed]
        assert len(pair) == 2, f"seed {seed} did not produce both seatings"
        diffs.append((pair[0] + pair[1]) / 2)

    arr = np.asarray(diffs)
    mean = float(arr.mean())
    ci95 = float(1.96 * arr.std(ddof=1) / np.sqrt(count)) if count > 1 else float("inf")
    return PairReport(count, mean, ci95, wins_a, wins_b, unfinished)

File: training-rs/src/evaluate.rs
//! Duplicate-deal paired evaluation: play two genomes head to head over a set
//! of deals, each deal in both seatings, and report the mean point differential
//! with a 95% CI.
//!
//! **Currently unused.** It computes the same paired differential that
//! [`crate::fitness`] now uses as the ES selection signal — the idea was
//! already here in Rust while the training loop was ranking on ELO. The two are
//! kept separate because they have different shapes: this owns a rollout and is
//! an A/B tool for two specific genomes; `fitness` is a pure fold over a whole
//! generation's results. `mean_diff` here and `fitness` there are in the same
//! unit (mean points per game), so they can be read against each other.

use crate::genome::{Arch, Genome};
use crate::policy::{forward_picks, WeightStack};
use crate::pool::Pool;
use candle_core::{DType, Device, Tensor};

pub struct PairReport {
    pub pairs: usize,
    pub mean_diff: f64,
    pub ci95: f64,
    pub wins_a: usize,
    pub wins_b: usize,
    pub unfinished: usize,
}

pub fn evaluate_pair(
    vec_a: &Genome,
    vec_b: &Genome,
    arch: &Arch,
    seeds: &[u64],
    device: &Device,
) -> PairReport {
    let count = seeds.len();
    let pool_seeds: Vec<u64> = seeds.iter().chain(seeds.iter()).copied().collect();
    let mut pool = Pool::new(pool_seeds, Some(200_000), None);

    let dtype = match device {
        Device::Cuda(_) => DType::BF16,
        _ => DType::F32,
    };
    let stack = WeightStack::from_roster(&[vec_a, vec_b], arch, device, dtype);

    while pool.has_live() {
        let encoded = pool.encode();
        let n_rows = encoded.rows.len();
        if n_rows == 0 {
            break;
        }
        let genome_idx: Vec<usize> = encoded
            .rows
            .iter()
            .map(|&(game_idx, seat)| {
                let a_is_team_zero = game_idx < count;
                if (seat % 2 == 0) == a_is_team_zero {
                    0
                } else {
                    1
                }
            })
            .collect();

        let obs = Tensor::from_vec(encoded.obs, (n_rows, arch.obs), device)
            .unwrap_or_else(|e| panic!("obs: {e}"));
        let acts = Tensor::from_vec(encoded.acts, (n_rows, encoded.width, arch.act), device)
            .unwrap_or_else(|e| panic!("acts: {e}"));
        let mask = Tensor::from_vec(
            encoded
                .mask
                .iter()
                .map(|&v| if v { 1u32 } else { 0u32 })
                .collect::<Vec<_>>(),
            (n_rows, encoded.width),
            device,
        )
        .unwrap_or_else(|e| panic!("mask: {e}"));

        let picks = forward_picks(&stack, &obs, &acts, &mask, &genome_idx);
        let _ = pool.apply(&picks);
    }

    let results = pool.results();
    let mut diffs = Vec::new();
    let mut wins_a = 0;
    let mut wins_b = 0;
    let mut unfinished = 0;
    let mut by_seed: std::collections::HashMap<u64, Vec<f64>> = std::collections::HashMap::new();

    for (index, (_seed, scores, winner, _hands, is_unfinished)) in results.iter().enumerate() {
        let a_team_zero = index < count;
        let a_score = if a_team_zero { scores[0] } else { scores[1] };
        let b_score = if a_team_zero { scores[1] } else { scores[0] };
        by_seed
            .entry(*_seed)
            .or_default()
            .push((a_score - b_score) as f64);
        if *is_unfinished {
            unfinished += 1;
        } else if let Some(w) = winner {
            let won_a = (*w == 0) == a_team_zero;
            if won_a {
                wins_a += 1;
            } else {
                wins_b += 1;
            }
        }
    }

    for seed in seeds {
        if let Some(pair) = by_seed.get(seed) {
            if pair.len() == 2 {
                diffs.push((pair[0] + pair[1]) / 2.0);
            }
        }
    }

    let mean = diffs.iter().sum::<f64>() / diffs.len().max(1) as f64;
    let variance = if diffs.len() > 1 {
        diffs.iter().map(|d| (d - mean).powi(2)).sum::<f64>() / (diffs.len() - 1) as f64
    } else {
        f64::INFINITY
    };
    let ci95 = 1.96 * variance.sqrt() / (count as f64).sqrt();

    PairReport {
        pairs: count,
        mean_diff: mean,
        ci95,
        wins_a,
        wins_b,
        unfinished,
    }
}
